cmd/collector/app/zipkin/zipkin_json_test.go

```diff
@@ -0,0 +1,390 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zipkin
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"math"
+	"testing"
+)
+
+var endpointFmt = `{"serviceName": "%s", "ipv4": "%s", "ipv6": "%s", "port": %d}`
+var annoFmt = `{"value": "%s", "timestamp": %d, "endpoint": %s}`
+var binaAnnoFmt = `{"key": "%s", "value": "%s", "endpoint": %s}`
+var spanFmt = `[{"name": "%s", "id": "%s", "parentId": "%s", "traceId": "%s", "timestamp": %d, "duration": %d, "debug": %t, "annotations": [%s], "binaryAnnotations": [%s]}]`
+
+func createEndpoint(serviveName string, ipv4 string, ipv6 string, port int) string {
+	return fmt.Sprintf(endpointFmt, serviveName, ipv4, ipv6, port)
+}
+
+func createAnno(val string, ts int, endpoint string) string {
+	return fmt.Sprintf(annoFmt, val, ts, endpoint)
+}
+
+func createBinAnno(key string, val string, endpoint string) string {
+	return fmt.Sprintf(binaAnnoFmt, key, val, endpoint)
+}
+
+func createSpan(name string, id string, parentID string, traceID string, ts int64, duration int64, debug bool,
+	anno string, binAnno string) string {
+	return fmt.Sprintf(spanFmt, name, id, parentID, traceID, ts, duration, debug, anno, binAnno)
+}
+
+func TestDecodeWrongJson(t *testing.T) {
+	spans, err := deserializeJSON([]byte(""))
+	require.Error(t, err)
+	assert.Nil(t, spans)
+}
+
+func TestUnmarshalEndpoint(t *testing.T) {
+	endp := &endpoint{}
+	err := json.Unmarshal([]byte(createEndpoint("foo", "127.0.0.1", "2001:db8::c001", 66)), endp)
+	require.NoError(t, err)
+	assert.Equal(t, "foo", endp.ServiceName)
+	assert.Equal(t, "127.0.0.1", endp.IPv4)
+	assert.Equal(t, "2001:db8::c001", endp.IPv6)
+	// only service name
```

Comment:
this test and a couple of other tests might be slightly more readable with fixtures